### PR TITLE
Fix replay export not working correctly from online leaderboards

### LINKED ISSUE #27859: Exporting a replay when viewing the score from within an online leaderboard causes it to be corrupted
### Type

Game behaviour

### Bug description

If you view a score from within an online leaderboard (Global, Country, Friends) and try to export the replay, it will export a corrupted 0KB file. Exporting it after watching it, or exporting it when viewing the score from within the "Local" leaderboard allows the replay to export without issue.

### Screenshots or videos

https://github.com/ppy/osu/assets/6813986/79b59ac2-58ef-4f65-8f9e-6a1503ae5004

### Version

2024.412.1

### Logs

[compressed-logs.zip](https://github.com/ppy/osu/files/14970808/compressed-logs.zip)

### PR DESCRIPTION
**Pull request by Joehuu**
- Fixes https://github.com/ppy/osu/issues/27859

Reuses some code from `PresentScore`.

### CODE REVIEW

**Review by bdach — CHANGES_REQUESTED** [review]

File: osu.Game/OsuGame.cs [hunk 1]
```diff
@@ -706,34 +706,17 @@ public void PresentScore(IScoreInfo score, ScorePresentType presentType = ScoreP
...
+            if (databasedScore == null) return;
 
             if (databasedScore.Replay == null)
             {
                 Logger.Log("The loaded score has no replay data.", LoggingTarget.Information);
                 return;
             }
 
-            var databasedBeatmap = BeatmapManager.QueryBeatmap(b => b.ID == databasedScoreInfo.BeatmapInfo.ID);
+            var databasedBeatmap = BeatmapManager.QueryBeatmap(b => b.OnlineID == score.Beatmap.OnlineID);
```
Comment (bdach) [inline osu.Game/OsuGame.cs:719]: What's happening here? Why was this changed?

File: osu.Game/Scoring/ScoreManager.cs [hunk 2]
```diff
@@ -60,18 +58,54 @@ public override bool PauseImports
             };
         }
 
-        public Score GetScore(ScoreInfo score) => scoreImporter.GetScore(score);
+        /// <summary>
+        /// Retrieve a <see cref="Score"/> from a given <see cref="IScoreInfo"/>.
+        /// </summary>
+        /// <param name="scoreInfo">The <see cref="IScoreInfo"/> to convert.</param>
+        /// <returns>The <see cref="Score"/>. Null if the score on the database cannot be found.</returns>
```
Comment (bdach) [inline osu.Game/Scoring/ScoreManager.cs:65]:
```suggestion
        /// <returns>The <see cref="Score"/>. Null if the score cannot be found in the database.</returns>
```

File: osu.Game/Scoring/ScoreManager.cs [hunk 3]
```diff
@@ -163,11 +197,25 @@ public override bool IsAvailableLocally(ScoreInfo model)
...
+        /// <remarks>
+        /// The <see cref="IScoreInfo"/> is re-retrieved from the database to ensure all the required data
+        /// for exporting a replay are present (may have missing properties if it was retrieved from online data).
+        /// </remarks>
+        public Task? Export(ScoreInfo scoreInfo)
+        {
+            ScoreInfo? databasedScoreInfo = getDatabasedScoreInfo(scoreInfo);
+
+            return databasedScoreInfo == null ? null : scoreExporter.ExportAsync(databasedScoreInfo.ToLive(Realm));
+        }
```
Comment (bdach) [inline osu.Game/Scoring/ScoreManager.cs:214]: `Task?` is a very awkward return type to work with (you can't `await` it directly without nullchecking first etc). The established practice is to either return `Task.CompletedTask` or just fail the task by throwing.

I don't particularly see any value in handling imports where the databased score cannot be found so I'd be probably okay with silently not exporting anything but something should probably be logged in such case at least. Maybe even at important level to let the user know their score didn't export. Throwing an exception may be fine too.

**Comment by bdach** [comment]
Code quality also needs work. Something like this probably.

```diff
diff --git a/osu.Game.Tests/Visual/Navigation/TestScenePresentScore.cs b/osu.Game.Tests/Visual/Navigation/TestScenePresentScore.cs
index 6590339311..004d1de116 100644
--- a/osu.Game.Tests/Visual/Navigation/TestScenePresentScore.cs
+++ b/osu.Game.Tests/Visual/Navigation/TestScenePresentScore.cs
@@ -170,7 +170,7 @@ private Func<ScoreInfo> importScore(int i, RulesetInfo ruleset = null)
                     BeatmapInfo = beatmap.Beatmaps.First(),
                     Ruleset = ruleset ?? new OsuRuleset().RulesetInfo,
                     User = new GuestUser(),
-                }).Value;
+                })!.Value;
             });
 
             AddAssert($"import {i} succeeded", () => imported != null);
diff --git a/osu.Game.Tests/Visual/UserInterface/TestSceneDeleteLocalScore.cs b/osu.Game.Tests/Visual/UserInterface/TestSceneDeleteLocalScore.cs
index 529874b71e..e2fe10fa74 100644
--- a/osu.Game.Tests/Visual/UserInterface/TestSceneDeleteLocalScore.cs
+++ b/osu.Game.Tests/Visual/UserInterface/TestSceneDeleteLocalScore.cs
@@ -104,7 +104,7 @@ protected override IReadOnlyDependencyContainer CreateChildDependencies(IReadOnl
                         Files = { new RealmNamedFileUsage(new RealmFile { Hash = $"{i}" }, string.Empty) }
                     };
 
-                    importedScores.Add(scoreManager.Import(score).Value);
+                    importedScores.Add(scoreManager.Import(score)!.Value);
                 }
             });
         });
diff --git a/osu.Game/Screens/Play/Player.cs b/osu.Game/Screens/Play/Player.cs
index 4fcc52bc5d..3a80caf259 100644
--- a/osu.Game/Screens/Play/Player.cs
+++ b/osu.Game/Screens/Play/Player.cs
@@ -1200,6 +1200,7 @@ protected virtual Task ImportScore(Score score)
             var importableScore = score.ScoreInfo.DeepClone();
 
             var imported = scoreManager.Import(importableScore, replayReader);
+            Debug.Assert(imported != null);
 
             imported.PerformRead(s =>
             {

```

File: osu.Game/OsuGame.cs (hunk 1 above)
Comment (Joehuu) [reply osu.Game/OsuGame.cs:719]: `databasedScoreInfo` was a `ScoreInfo` and there is no `BeatmapInfo` in `IScoreInfo`.

Can't really bring back `databasedScoreInfo` without duplicating `getDatabasedScoreInfo()` locally or making it public (which is not wanted).

Seems to work in limited testing, but I wasn't too confident with this line.

Comment (bdach) [reply osu.Game/OsuGame.cs:719]:
```suggestion
            var databasedBeatmap = BeatmapManager.QueryBeatmap(b => b.ID == databasedScore.ScoreInfo.BeatmapInfo.ID);
```

surely?

**Review by bdach — APPROVED** [review]
